modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/DefaultExtensionManager.java

```diff
@@ -0,0 +1,425 @@
+            ConfigurationInstanceProvider<?> instanceProvider = instanceProviderEntry.getValue();
+            registerConfigurationInstanceProvider(instanceProviderEntry.getKey(), instanceProvider);
+        }
+    }
+
+    @Override
+    public List<Extension> discoverExtensions(ClassLoader classLoader)
+    {
+        LOGGER.info("Starting discovery of extensions");
+
+        List<Extension> discovered = extensionDiscoverer.discover(classLoader);
+        LOGGER.info("Discovered {} extensions", discovered.size());
+
+        ImmutableList.Builder<Extension> accepted = ImmutableList.builder();
+
+        for (Extension extension : discovered)
+        {
+            if (registerExtension(extension))
+            {
+                accepted.add(extension);
+            }
+        }
+
+        return accepted.build();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean registerExtension(Extension extension)
+    {
+        LOGGER.info("Registering extension {} (version {})", extension.getName(), extension.getVersion());
+        final String extensionName = extension.getName();
+
+        if (register.containsExtension(extensionName))
+        {
+            return maybeUpdateExtension(extension, extensionName);
+        }
+        else
+        {
+            doRegisterExtension(extension, extensionName);
+            return true;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public <C> void registerConfigurationInstanceProvider(String providerName, ConfigurationInstanceProvider<C> configurationInstanceProvider)
+    {
+        Configuration configuration = configurationInstanceProvider.getConfiguration();
+        ExtensionStateTracker extensionStateTracker = register.getExtensionState(configuration);
+        extensionStateTracker.registerConfigurationInstanceProvider(configuration, providerName, configurationInstanceProvider);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public <C> C getConfigurationInstance(final ConfigurationInstanceProvider<C> configurationInstanceProvider, OperationContext operationContext)
+    {
+        return configurationInstanceProvider.get(operationContext, new ConfigurationInstanceRegistrationCallback()
+        {
+            @Override
+            public <C> void registerNewConfigurationInstance(ConfigurationInstanceProvider<C> configurationInstanceProvider, C configurationInstance)
+            {
+                registerConfigurationInstance(configurationInstanceProvider.getConfiguration(),
+                                              configurationInstanceProvider.getName(),
+                                              configurationInstance);
+            }
+        });
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public OperationExecutor getOperationExecutor(OperationContext operationContext)
+    {
+        Extension extension = register.getExtension(operationContext.getOperation());
```

Comment:
I had to follow many methods in order to understand this, it was not intuitive for me